### PR TITLE
[Ready]Sticky ban in game interface

### PR DESCRIPTION
**Pull request by MrsTonedOne**
This allows admins to add and remove dream deamon sticky bans.

Minor credit to /vg/ as an example for the primary SetConfig() code

This marks the stickybans it adds so that it can only be used to edit/remove sticky bans it created.
- [x] Viewing sticky bans
- [x] Adding sticky bans
- [x] Editing sticky bans
- [x] Removing sticky bans
- [x] Delinking ckeys from sticky bans
- [x] Edit isbanned() to better handle deamon bans
- [ ] Exempting ckeys from matching a given sticky ban
- [x] Making admins immuned to stickybans but only if placed by the code
- [x] Clean up code
- [ ] Make the interface less shit.
- [x] ~~Add a prompt for making a perma ban a sticky ban when banning (how /vg/ does it)~~ (decided against this for now)

Close enough.

Other notes:

isbanned() cleaned up, it logs sticky ban matches, and better handles admins being exempt from bans
Adminbans now still work once the admin is demoted.
Admins bypassing a ban because they are an admin is logged and announced to all admins including the one who walked past it.
Admins are now exempt from host bans. (this only applies to host bans for ss13, global host bans (where the 'apply to this game only' checkbox is not checked (defaults to not checked)) do not trigger isbanned() and thus, admins can not bypass them, no matter what we do.)

Added a system to queue a message for a client, to be shown next time they connect, this was needed because isbanned() is called before the client is created, so if you want to send a message to an admin, letting them know they just walked pass a matching ban, you have to do it this way.

### CODE REVIEW

**Comment by Core0verload** [comment]
How BYOND's internal ban system works?

**Comment by SconesC** [comment]
Fucking awesome

**Comment by MrStonedOne** [comment]
>How BYOND's internal ban system works?

Stupidly robust. Sticky bans (dream deamon bans with the stickyban checkbox checked) have stopped most ban evaders who use vms or the cid randomizer dll.

The most notable example is terbs. We added a sticky ban on his original account that had been banned for over a year, after he came in to irc bragging about how he was gonna ban evade now that he moved and got a new computer, isp, and everything...

Think about that, new isp, new computer, new location, no trace right?

Sticky ban caught him trying evade under the username purity ring.

Its fucking black magic.

**Comment by Miauw62** [comment]
all the turds in the byond codebase probably compressed it into a black hole of coding, and sticky bans happened to be at the center.

**Comment by Core0verload** [comment]
I suspect some sort of combined hardware/software/provider fingerprinting, including embedded IE parameters fingerprinting, because I tried to use embedded IE fingerprinting at least once and it mostly worked. Will test it on local server anyway. Hope these bans are not *too* sticky.

**Comment by freerealestate** [comment]
oh i thought this was an interface to ban sticky

maybe next time

**Comment by Jalleo** [comment]
He probably used that account on his old computer but elsewhere in byond.
Also there needs to be sanity checks with this mso funny thimg happened a couple o weeks ago due to a badly setup sticky ban.

**Comment by MrStonedOne** [comment]
>Also there needs to be sanity checks with this mso funny thimg happened a couple o weeks ago due to a badly setup sticky ban.

I know about that, and already accounted for them

>maybe he just had more than one computer when he got the new one and connected with the alternative one thinking he hadn't connected to the server with it in the past?

<SplitQuote>

>He probably used that account on his old computer but elsewhere in byond.

Terbs knows what he's doing, i doubt he'd fuck up like that.

**Comment by Chnkr** [comment]
> Its fucking black magic.
Not really, i just tested how sticky bans actually work on localhost and i could evade them pretty easily without any effort, it just requires a extra step when registering a new account.

**Comment by Jalleo** [comment]
Welp our secret weapon aint so poweful anymore then.

**Comment by optimumtact** [comment]
hello

**Comment by Razharas** [comment]
@MrStonedOne

**Comment by MrStonedOne** [comment]
I need to get back on this, I kept getting stuck at reworking the logic of isbanned() so I think I'll scrap a few features for now to make it manageable.

**Comment by Razharas** [comment]
Weekly ping of activity reminder
@MrStonedOne

**Comment by Niknakflak** [comment]
Hoping this gets finished soon @MrStonedOne

**Comment by tedward1337** [comment]
@mrstonedone tick tock

**Comment by lollerderby** [comment]
Great work @MrStonedOne I can't wait to see this implemented @MrStonedOne.

**Comment by LanCartwright** [comment]
I want this.

**Comment by Niknakflak** [comment]
@mrstonedone

**Comment by Niknakflak** [comment]
@mrstonedone

**Comment by Jalleo** [comment]
Just give it time to mature and be ready treat it like a whiskey.

**Comment by Razharas** [comment]
I want to still be young when this is ready to merge so
@MrStonedOne

**Comment by Shadowlight213** [comment]
@Mrstonedone

**Comment by duncathan** [comment]
@MrStonedOne

File: code/modules/admin/IsBanned.dm [hunk 1]
```diff
@@ -63,10 +28,13 @@
...
 		. = CheckBan( ckey(key), computer_id, address )
 		if(.)
-			log_access("Failed Login: [key] [computer_id] [address] - Banned [.["reason"]]")
-			return .
-
-		return ..()	//default pager ban stuff
+			if (admin)
+				log_admin("The admin [key] has been allowed to bypass a matching ban on [.["key"]]")
+				message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching ban on [.["key"]]</span>")
+				addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching ban on [.["key"]]")
```
Comment (Markolie) [inline code/modules/admin/IsBanned.dm]: Missing </span> here.

File: code/modules/admin/IsBanned.dm [hunk 2]
```diff
@@ -99,12 +67,46 @@
...
+				if (pckey != ckey)
+					continue
+			if (admin)
+				if (bantype == "ADMIN_PERMABAN" || bantype == "ADMIN_TEMPBAN")
+					log_admin("The admin [key] is admin banned, and has been disallowed access")
+					message_admins("<span class='adminnotice'>The admin [key] is admin banned, and has been disallowed access</span>")
+				else
+					log_admin("The admin [key] has been allowed to bypass a matching ban on [pckey]")
+					message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching ban on [pckey]</span>")
+					addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching ban on [pckey]")
```
Comment (Markolie) [inline code/modules/admin/IsBanned.dm]: Closing span

File: code/modules/admin/IsBanned.dm [hunk 3]
```diff
@@ -99,12 +67,46 @@
...
+
+	. = ..()	//default pager ban stuff
+	if (.)
+		//byond will not trigger isbanned() for "global" host bans,
+		//ie, ones where the "apply to this game only" checkbox is not checked (defaults to not checked)
+		//So it's safe to let admins walk thru host/sticky bans here
+		if (admin)
+			log_admin("The admin [key] has been allowed to bypass a matching host/sticky ban")
+			message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching host/sticky ban</span>")
+			addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching host/sticky ban")
```
Comment (Markolie) [inline code/modules/admin/IsBanned.dm]: Closing span

**Comment by Razharas** [comment]
@MrStonedOne 
So is this ready or are you going to achieve some more goals from op post?

**Comment by MrStonedOne** [comment]
@Razharas This is as ready as it will ever be. The interface is clunky, but I suck at simple html interfaces, so it will just have to do.

**Comment by Razharas** [comment]
Ill pretend i reviewed it so it can be tested on live servs for actual problem finding

**Comment by d3athrow** [comment]
once you know how a sticky ban works it's trivial to bypass, but most people will never figure out how.

**Pull request merged**